src/ipc/ipc4/helper.c

```diff
@@ -331,6 +342,343 @@ int ipc_comp_disconnect(struct ipc *ipc, ipc_pipe_comp_connect *_connect)
 	return 0;
 }
 
+/* dma index may be for playback or capture. Current
+ * hw supports PLATFORM_MAX_DMA_CHAN of playback
+ * and the index more than this will be capture. This function
+ * convert dma id to dma channel
+ */
+static inline int process_dma_index(uint32_t dma_id, uint32_t *dir, uint32_t *chan)
+{
+	if (dma_id > DAI_NUM_HDA_OUT + DAI_NUM_HDA_IN)
+		return -EINVAL;
+
+	if (dma_id >= PLATFORM_MAX_DMA_CHAN) {
+		*dir = SOF_IPC_STREAM_CAPTURE;
+		*chan = dma_id - PLATFORM_MAX_DMA_CHAN;
+	} else {
+		*dir = SOF_IPC_STREAM_PLAYBACK;
+		*chan = dma_id;
+	}
+
+	return 0;
+}
+
+static struct comp_dev *ipc4_create_host(uint32_t pipeline_id, uint32_t id, uint32_t dir)
+{
+	struct sof_uuid uuid = {0x8b9d100c, 0x6d78, 0x418f, {0x90, 0xa3, 0xe0,
+			0xe8, 0x05, 0xd0, 0x85, 0x2b}};
+	struct ipc_config_host ipc_host;
+	struct comp_ipc_config config;
+	const struct comp_driver *drv;
+	struct comp_dev *dev;
+
+	drv = ipc4_get_drv((uint8_t *)&uuid);
+	if (!drv)
+		return NULL;
+
+	memset_s(&config, sizeof(config), 0, sizeof(config));
+	config.type = SOF_COMP_HOST;
+	config.pipeline_id = pipeline_id;
+	config.core = 0;
+	config.id = id;
+
+	memset(&ipc_host, 0, sizeof(ipc_host));
+	ipc_host.direction = dir;
+
+	dev = drv->ops.create(drv, &config, &ipc_host);
+	if (!dev)
+		return NULL;
+
+	dev->direction = dir;
+	list_init(&dev->bsource_list);
+	list_init(&dev->bsink_list);
+
+	return dev;
+}
+
+static struct comp_dev *ipc4_create_dai(struct pipeline *pipe, uint32_t id, uint32_t dir,
+					struct ipc4_copier_module_cfg *copier_cfg)
+{
+	struct sof_uuid uuid = {0xc2b00d27, 0xffbc, 0x4150, {0xa5, 0x1a, 0x24,
+				0x5c, 0x79, 0xc5, 0xe5, 0x4b}};
```

Comment:
ditto ? we need to comment this. I assume its because chain DMA uses not topology ?

Reply:
sure, No topology for chain DMA. Driver only create a chain DMA pipeline by a simple ipc message with host dma id and link dma id